File: Dec1/puzzle2/solution.py
def read_data(file_name):
    """Read data from file"""
    _data = []
    with open(file_name, "r") as file:
        for line in file:
            _data.append(
                tuple(map(int, line.split()))
            )
    return _data


if __name__ == "__main__":
    data = read_data("data.txt")

    n = len(data)
    score = 0
    l1 = [x for x, y in data]
    l2 = [y for x, y in data]

    # create a hash map for the counts in list 2: O(n)
    count = {}
    for i in range(n):
        if l2[i] not in count:
            count[l2[i]] = 1
        else:
            count[l2[i]] += 1
    # multiple count with number in list 1: O(n)
    for i in range(n):
        score += l1[i] * count.get(l1[i], 0)

    print(score)

File: Dec1/puzzle2/solution.ml
(* Read data from a file *)
let read_data file_name =
  let ic = open_in file_name in
  let rec loop acc =
    try
      let line = input_line ic in
      let nums = List.map int_of_string (String.split_on_char ' ' (String.trim line)) in
      match nums with
      | [x; y] -> loop ((x, y) :: acc)
      | _ -> failwith "Malformed line in file"
    with End_of_file ->
      close_in ic;
      List.rev acc
  in
  loop []

(* Main program *)
let () =
  let data = read_data "Dec1/puzzle2/data.txt" in
  let n = List.length data in

  (* Tipp: fst gives first element and snd the second *)
  let l1 = List.map fst data in
  let l2 = List.map snd data in
  
  (* similiar to python solution keep track of count through a has table *)
  let count = Hashtbl.create n in
  List.iter (fun y ->
    Hashtbl.replace count y ((Hashtbl.find_opt count y |> Option.value ~default:0) + 1)
  ) l2;

  (* Compute the score *)
  let score =
    List.fold_left (fun acc x ->
      acc + x * (Hashtbl.find_opt count x |> Option.value ~default:0)
    ) 0 l1
  in

  Printf.printf "%d\n" score